tests/Process/ComposerFinderTest.php

```diff
@@ -0,0 +1,108 @@
+<?php
+
+namespace PhpTuf\ComposerStager\Tests\Process;
+
+use PhpTuf\ComposerStager\Exception\FileNotFoundException;
+use PhpTuf\ComposerStager\Process\ComposerFinder;
+use PhpTuf\ComposerStager\Tests\TestCase;
+use Symfony\Component\Process\ExecutableFinder;
+
+/**
+ * @coversDefaultClass \PhpTuf\ComposerStager\Process\ComposerFinder
+ * @covers \PhpTuf\ComposerStager\Process\ComposerFinder::__construct
+ * @covers \PhpTuf\ComposerStager\Process\ComposerFinder::find
+ * @uses \PhpTuf\ComposerStager\Exception\FileNotFoundException
+ * @uses \PhpTuf\ComposerStager\Exception\PathException
+ *
+ * @property \Prophecy\Prophecy\ObjectProphecy|\Symfony\Component\Process\ExecutableFinder $executableFinder
+ */
+class ComposerFinderTest extends TestCase
+{
+    protected function setUp(): void
+    {
+        $this->executableFinder = $this->prophesize(ExecutableFinder::class);
+        $this->executableFinder
+            ->find('composer')
+            ->willReturnArgument();
+    }
+
+    private function createSut(): ComposerFinder
+    {
+        $executableFinder = $this->executableFinder->reveal();
+        return new ComposerFinder($executableFinder);
+    }
+
+    /**
+     * @dataProvider providerFind
+     */
+    public function testFind($path): void
+    {
+        $this->executableFinder
+            ->addSuffix('.phar')
+            ->shouldBeCalledOnce()
+            ->willReturn(null);
+        $this->executableFinder
+            ->find('composer')
+            ->shouldBeCalledOnce()
+            ->willReturn($path);
+
+        $sut = $this->createSut();
+
+        $expected = $sut->find();
+        // Call again to test result caching.
+        $sut->find();
+
+        self::assertSame($path, $expected, 'Returned correct path');
+    }
+
+    public function providerFind(): array
+    {
+        return [
+            ['/lorem/ipsum'],
+            ['/dolor/sit'],
+        ];
+    }
+
+    public function testFindNotFound(): void
+    {
+        $this->expectException(FileNotFoundException::class);
+        $this->expectExceptionMessageMatches('/found/');
+        $this->executableFinder
+            ->addSuffix('.phar')
+            ->shouldBeCalledOnce()
+            ->willReturn(null);
+        $this->executableFinder
+            ->find('composer')
+            ->shouldBeCalledOnce()
+            ->willReturn(null);
+
+        $sut = $this->createSut();
+
+        $sut->find();
+    }
+
+    /**
+     * Make sure ::find caches result when Composer is not found.
+     */
+    public function testFindNotFoundCaching(): void
+    {
+        $this->executableFinder
+            ->addSuffix('.phar')
+            ->shouldBeCalledOnce()
+            ->willReturn(null);
+        $this->executableFinder
+            ->find('composer')
+            ->shouldBeCalledOnce()
+            ->willReturn(null);
+        $sut = $this->createSut();
+
+        try {
+            $sut->find();
```

Comment:
In order for these to actually fail if the exception isn't thrown, we really should be doing something like this:

```
try {
  $sut->find();
  $this->fail('Expected a FileNotFoundException but nothing materialized');
}
catch (....)
```

Reply:
Actually, @phenaproxima, it already _does_ fail in that case, because the expectation above (line 95) asserts that it doesn't call `ComposerFinder::find()` again--which is the essence of what I'm asserting here, i.e., that we don't hit the filesystem more than once. I tested when writing it (red/green/refactor, you know), and that is demonstrably the case.